src/Tools/dotnet-user-secrets/src/Internal/ProjectIdResolver.cs

```diff
@@ -103,13 +103,13 @@ public string Resolve(string project, string configuration)
 
             if (!File.Exists(outputFile))
             {
-                throw new InvalidOperationException(Resources.FormatError_ProjectMissingId(projectFile));
+                throw new FileNotFoundException(Resources.FormatError_ProjectMissingId(projectFile));
             }
 
             var id = File.ReadAllText(outputFile)?.Trim();
             if (string.IsNullOrEmpty(id))
             {
-                throw new InvalidOperationException(Resources.FormatError_ProjectMissingId(projectFile));
+                throw new NullReferenceException(Resources.FormatError_ProjectMissingId(projectFile));
```

Comment:
I didn't even notice the `catch (NullReferenceException)`! Catching either an `InvalidOperationException` or a `NullReferenceException` is a code smell. A `NullReferenceException` should never happen in any code we ship no matter how it's called. An `InvalidOperationException` much like an `ArgumentException` should be avoidable if the caller correctly checks state and input beforehand to ensure the operation is valid.

If checking whether the operation is valid beforehand is hard or impossible to do, an Exception still should not be used for control flow. I get this tool doesn't need to be super high performance or anything, but I think `Resolve(string project, string configuration)` should returning a `string?` instead of a `string` for the secrets ID is cleaner.

Do we have any tests that hit this code path and verify we generate a secrets ID if you do a user-jwts command without creating one first?

Reply:
> Do we have any tests that hit this code path and verify we generate a secrets ID if you do a user-jwts command without creating one first?

`Create_WarnsOnNoSecretInProject` does this.

> I get this tool doesn't need to be super high performance or anything, but I think Resolve(string project, string configuration) should returning a string? instead of a string for the secrets ID is cleaner.

Yeah, I suppose instead of throwing exceptions we can just report. I was actually a little surprised that this tool threw user-facing exceptions in the first place because that feels like an experience smell.

Reply:
Changed this up. This API no longer throws and instead logs exceptions and returns `string?`.